reazon.el

```diff
@@ -425,10 +425,12 @@ Also known as committed choice. This operator is impure."
   "Define relation NAME with args VARLIST and body GOALS."
   (declare (indent 2))
   (let ((stream (gensym)))
-    `(defun ,name ,varlist
-       (lambda (,stream)
-         (lambda ()
-           (funcall (reazon-conj ,@goals) ,stream))))))
+    `(if (not lexical-binding)
+         (user-error "Can't define new relations when lexical binding is disabled!")
+       (defun ,name ,varlist
+         (lambda (,stream)
+           (lambda ()
+             (funcall (reazon-conj ,@goals) ,stream)))))))
```

Comment:
moved the lexical check into the macro expansion. for some reason the byte-compiler really doesn't like my change:

```lang
make compile
emacs -Q --batch --directory ./ --funcall batch-byte-compile .//reazon-perf.el .//reazon-sudoku.el .//reazon.el
emacs -Q --batch --directory ./ --directory ./test --funcall batch-byte-compile ./test/reazon-test-interface.el ./test/reazon-test-prolog.el ./test/reazon-test-puzzle.el ./test/reazon-test-relation.el ./test/reazon-test-sudoku.el ./test/reazon-test-trs.el ./test/reazon-test-unit.el ./test/reazon-test-utils.el

In toplevel form:
test/reazon-test-interface.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-prolog.el:27:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-puzzle.el:27:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-relation.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-sudoku.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-trs.el:29:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-unit.el:30:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-utils.el:31:1:Error: Can’t define new relations when lexical binding is disabled!
make: *** [compile] Error 1
```

i really have no clue why this is happening!

Reply:
Somehow these definitions are getting evaluated outside of lexical scoping. But where? `reazon.el` has lexical scoping and so does `bytecomp.el`. I don't understand who's in control of the scoping when batch compilation is run.